### ISSUE TITLE
crash when missing SCHEDULE_EXACT_ALARM permission and using androidAllowWhileIdle

### ISSUE THREAD

**Issue by armandojimenez** [MaikuB/flutter_local_notifications#1526]
Hi, I am getting this error in crashlytics:
```
Fatal Exception: java.lang.RuntimeException: Unable to start receiver com.dexterous.flutterlocalnotifications.ScheduledNotificationBootReceiver: java.lang.SecurityException: Caller dev.armandojimenez.believe needs to hold android.permission.SCHEDULE_EXACT_ALARM to set exact alarms.
       at android.app.ActivityThread.handleReceiver(ActivityThread.java:4757)
       at android.app.ActivityThread.access$1800(ActivityThread.java:309)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2279)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8582)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:563)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1133)
```
Looks like in android 12, if we use androidAllowWhileIdle : true, we need to add the SCHEDULE_EXACT_ALARM permission, or set androidAllowWhileIdle to false. This causes a crash if the user revoke the permission via system settings.

**Expected behavior**
No crash when using androidAllowWhileIdle param

**Sample code to reproduce the problem**
Use androidAllowWhileIdle : true, and then try to set the notification without the permission in the manifest or if the user revoke the permission causes an exception [Example article](https://www.droidcon.com/2021/11/30/quick-guide-of-alarms-on-android-12/)

**Comment by MaikuB** [comment]
Provided you are on a more up to date version of the plugin with Android 12 support, this permission would've been requested by the plugin itself https://github.com/MaikuB/flutter_local_notifications/blob/2c0b2db29fef939f1355cd708f8c5fee73deed22/flutter_local_notifications/android/src/main/AndroidManifest.xml#L6

Note that it falls to your application to check if the appropriate permission has been granted in case like the article covers. Have you done this? If not, then this is something you'd need to address This would not be the responsibility of the plugin as it's up to each app to decide what it should do when permissions haven't been granted or revoked e.g. present UI, redirect to settings etc.

**Comment by armandojimenez** [comment]
Ok I see, thanks. Would it be good to add a check to see if we have the permission to set an exact alarm? and a way to send the user to the permission screen. Or do you think I should request this in the permission_handler package?

Also, this info should be added to the readme, since the package will crash if the user remove the permission from the system settings.

**Comment by MaikuB** [comment]
> Would it be good to add a check to see if we have the permission to set an exact alarm? and a way to send the user to the permission screen. Or do you think I should request this in the permission_handler package?

If you mean if *all* of what you mentioned should be part of a plugin then no as any kind of redirection is specific to each app's use case. Permissions checking though is better suited to a permissions plugin.

> Also, this info should be added to the readme, since the package will crash if the user remove the permission from the system settings.

Exactly which version of the plugin was your app using though?

**Comment by armandojimenez** [comment]
Version 9.3.2

**Comment by MaikuB** [comment]
Do you have exception handling around the code in your own app around when a notification is going to be scheduled? The reason I'm asking is because the plugin throwing an exception here makes sense and mirrors what would happen if you to write all of your app entirely with Java or Kotlin. From what I've also seen when this results in a [PlatformException](https://api.flutter.dev/flutter/services/PlatformException-class.html) and therefore could be caught by your application. Have you checked that this occurs?

**Comment by tigertore** [comment]
The scheduling of alarms in flutter_local_notifications on Android is using 

AlarmManagerCompat.setExact()
AlarmManagerCompat.setExactAndAllowWhileIdle()

Both methods require the android.permission.SCHEDULE_EXACT_ALARM on Android 12.

If the user revokes this permission we can't schedule any future notifications.

The SCHEDULE_EXACT_ALARM is about saving battery and prevent apps from running in background. 

I would prefer to schedule a non-exact alarm instead, if the user revoked this permission.
AlarmManger.set()
AlarmManger.setAndAllowWhileIdle()

Have you concidered adding support for that ?

**Comment by MaikuB** [comment]
@tigertore no, as i view that as being an exception case i.e. one that is more specific to your app that should have its own logic to do so. Developers typically pick a cross-platform library and develop cross-platform apps to achieve feature parity so using inexact alarms wouldn't achieve this. You should find other plugins in other ecosystems using exact alarms for similar reasons.

Exact alarms consuming more battery has always been the case before, it's just that Android 12 had a permission to control this and using exact alarms for notifications is also an acceptable use case per the Android docs (see https://developer.android.com/training/scheduling/alarms#exact-acceptable-use-cases). The fact that it stops working should be expected much like how you wouldn't be able to show a user's location if they revoked that permission later. Given how this permission is explained to users in Android settings (i.e. an app can use this to schedule actions like alarms or reminders), if a user turns this off then the fact that you want to schedule a notification through other means (i.e. inexact alarms) can be viewed as not respecting the user's intent and cause more confusion. 

The Android documentation also states that it's up to apps to explain why the permission is needed too and redirect users to enable the permission if required. This is IMO the better way to deal with this as this is also how it's commonly dealt with for other permissions, particularly when you factor what I wrote earlier on how the exact alarm permission is explained to users. 
 
Furthermore, it sounds you want to fallback to inexact alarms immediately after the permission has been revoked and that is not possible.  The [ACTION_SCHEDULE_EXACT_ALARM_PERMISSION_STATE_CHANGED](https://developer.android.com/reference/android/app/AlarmManager#ACTION_SCHEDULE_EXACT_ALARM_PERMISSION_STATE_CHANGED) is only sent when an app is granted permission. Now this is something that is missing from the plugin where it could try to reschedule notifications after the permission has been regranted

**Comment by youssefali424** [comment]
" no, as i view that as being an exception case i.e. one that is more specific to your app that should have its own logic to do so"
I really appreciate your efforts for doing this for the community and even having the time to respond to us.

it is mentioned here that we can check if the permission is given or not https://developer.android.com/training/scheduling/alarms#exact-permission-check
the current behavior crashes the app, and it doesn't give a [PlatformException](https://api.flutter.dev/flutter/services/PlatformException-class.html) and noticed it as well on Crashlytics
i think it can be added to the plugin and won't be over the requirements for the plugin don't you think so?

and it will be safer to do this now as after Android 12 I think the same behaviour would remain the same

**Comment by ArmandoJimenezOS** [comment]
Also, would it be posible to do something like check if the permission is granted, if not then use setAndAllowWhileIdle() so at least the app does not crash and the notification is delivered (Not at the exact expected time)?

I think this plugin needs 2 permissions right now to work currently in Android, and the plugin offers a check or method for one of them (Notifications permission). Maybe we can pass a param we can pass so the plugin checks for exact alarm permission and fallback to setAndAllowWhileIdle() instead of setExactAndAllowWhileIdle()? Or we can check on our side with a permission handler, and pass a param so the plugin schedule and inexact alarm.

I ask this because if the user does not have the permission, we need to redirect them to the "special app access", and that screen is not easy to explain what to do for the typical user, or why the users need to do that.

**Comment by holocronweaver** [comment]
I would strongly prefer that a non-exact scheduled notification fallback be optional and disabled by default. In most cases developers scheduling a notification expect it to fire roughly at that time - doing otherwise would violate the principle of least surprise and thus should require an explicit parameter in the plugin API (maybe something like `fallbackToInexact`). Adding a separate inexact API to the plugin is also option.

**Comment by MaikuB** [comment]
Closing this now as the PR to add support for inexact notifications was released in 14.0.0. That PR also made changes to check the status of the exact alarm permission and should it find it's been revoked then will write an error log message as opposed to throwing an exception
